File: frontend/src/types/job.ts
export type JobStatusLiteral = "pending" | "running" | "completed" | "failed";

export interface JobProgress {
  current: number;
  total: number;
  percent: number;
}

export interface JobStatus {
  job_id: string;
  status: JobStatusLiteral;
  progress: JobProgress;
  message: string;
  result: Record<string, unknown> | null;
  error: string | null;
  metrics: Record<string, number>[];
  device_used: string | null;
  batch_progress: {
    epoch: number;
    total_epochs: number;
    batch: number;
    total_batches: number;
    images_done: number;
    images_total: number;
    speed_it_s: number;
    losses: Record<string, number>;
  } | null;
}

File: backend/app/schemas/job.py
from typing import Any, Literal

from pydantic import BaseModel, Field

JobStatusLiteral = Literal["pending", "running", "completed", "failed"]


class JobProgress(BaseModel):
    current: int = 0
    total: int = 0
    percent: float = 0.0


class JobStatus(BaseModel):
    job_id: str
    status: JobStatusLiteral = "pending"
    progress: JobProgress = JobProgress()
    message: str = ""
    result: dict[str, Any] | None = None
    error: str | None = None
    # Per-epoch training metrics, appended incrementally so the frontend can chart them live.
    metrics: list[dict[str, Any]] = Field(default_factory=list)
    # The actual torch device the training run resolved to (e.g. "cuda:0"), set once training
    # actually starts, so the UI can confirm the GPU/CPU choice was really honored.
    device_used: str | None = None
    # Within-epoch batch progress, updated every batch so the UI can mirror the terminal's
    # live tqdm bar (batch count, images processed, iterations/sec).
    batch_progress: dict[str, Any] | None = None
